### PR TITLE
feat: change kernel parameter handling

### PR DESCRIPTION
**Pull request by vmx**
BREAKING CHANGE: Kernel creation takes different parameter values now.

Defining the global work size now follows the CUDA model and no longer
the OpenCL model. In OpenCL the global work size is the total number
of threads. In CUDA it is the number of thread groups of local work
size.

This change is made, so that the same kernel creation code can be run
indepent of whether OpenCL or CUDA is used.

### CODE REVIEW

File: src/opencl/error.rs [hunk 1]
```diff
@@ -20,6 +20,8 @@ pub enum GPUError {
     IO(#[from] std::io::Error),
     #[error("Vendor {0} is not supported.")]
     Vendor(String),
+    #[error("{0}")]
+    ParseId(String),
```
Comment (dignifiedquire) [inline src/opencl/error.rs:24]: maybe better `InvalidId`?

File: src/opencl/mod.rs [hunk 2]
```diff
@@ -18,14 +19,144 @@ use opencl3::memory::CL_MEM_READ_WRITE;
 use opencl3::program::ProgramInfo::CL_PROGRAM_BINARIES;
 use opencl3::types::CL_BLOCKING;
 
-pub type BusId = u32;
+pub const CL_UUID_SIZE_KHR: usize = 16;
```
Comment (dignifiedquire) [inline src/opencl/mod.rs:22]: can you add a comment what this exactly is?

File: src/opencl/mod.rs [hunk 3]
```diff
@@ -18,14 +19,144 @@ use opencl3::memory::CL_MEM_READ_WRITE;
...
 
 const AMD_DEVICE_VENDOR_STRING: &str = "AMD";
 const NVIDIA_DEVICE_VENDOR_STRING: &str = "NVIDIA Corporation";
 
 #[allow(non_camel_case_types)]
 pub type cl_device_id = opencl3::types::cl_device_id;
 
+// The PCI-ID is the combination of the PCI Bus ID and PCI Device ID.
+///
+/// It is the first two identifiers of e.g. `lcpci`:
```
Comment (dignifiedquire) [inline src/opencl/mod.rs:32]: typo, should be `lspci`

File: src/opencl/mod.rs [hunk 4]
```diff
@@ -18,14 +19,144 @@ use opencl3::memory::CL_MEM_READ_WRITE;
...
+}
+
+/// Formats the UUID the same way as `clinfo` does, as an example:
+/// the output should looks like 46abccd6-022e-b783-572d-833f7104d05f
+impl fmt::Display for DeviceUuid {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}-{}-{}-{}-{}",
+            hex::encode(&self.0[..4]),
```
Comment (dignifiedquire) [inline src/opencl/mod.rs:117]: does this do padding? It looks like those UUIDs are padded in some form

File: src/opencl/mod.rs [hunk 5]
```diff
@@ -64,19 +195,21 @@ pub struct Device {
...
-        self.bus_id.hash(state);
+        self.pci_id.hash(state);
+        self.uuid.hash(state);
     }
 }
 
 impl PartialEq for Device {
     fn eq(&self, other: &Self) -> bool {
-        self.bus_id == other.bus_id
+        self.pci_id == other.pci_id && self.uuid == other.uuid
```
Comment (dignifiedquire) [inline src/opencl/mod.rs:212]: shouldn't the vendor and the name at least be checked as well?

File: src/opencl/mod.rs [hunk 6]
```diff
@@ -100,81 +233,57 @@ impl Device {
...
-        }
-    }
-}
-
-fn get_device_bus_id_by_index(index: usize) -> Option<BusId> {
-    if let Some(device) = get_device_by_index(index) {
-        device.bus_id
-    } else {
-        None
+    pub fn cl_device_id(&self) -> cl_device_id {
```
Comment (dignifiedquire) [inline src/opencl/mod.rs:282]: comment?

File: src/opencl/mod.rs [hunk 7]
```diff
@@ -100,81 +233,57 @@ impl Device {
...
+            .find(|d| match d.uuid {
+                Some(id) => uuid == id,
                 None => false,
             })
             .ok_or(GPUError::DeviceNotFound)
     }
 
-    pub fn cl_device_id(&self) -> cl_device_id {
-        self.device.id()
+    pub fn by_unique_id(unique_id: UniqueId) -> GPUResult<&'static Device> {
```
Comment (dignifiedquire) [inline src/opencl/mod.rs:272]: comment

File: src/opencl/mod.rs [hunk 8]
```diff
@@ -100,81 +233,57 @@ impl Device {
...
-    pub fn by_bus_id(bus_id: BusId) -> GPUResult<&'static Device> {
+    pub fn by_pci_id(pci_id: PciId) -> GPUResult<&'static Device> {
         Self::all_iter()
-            .find(|d| match d.bus_id {
-                Some(id) => bus_id == id,
+            .find(|d| pci_id == d.pci_id)
+            .ok_or(GPUError::DeviceNotFound)
+    }
+
+    pub fn by_uuid(uuid: DeviceUuid) -> GPUResult<&'static Device> {
```
Comment (dignifiedquire) [inline src/opencl/mod.rs:263]: comment

File: src/opencl/mod.rs [hunk 9]
```diff
@@ -100,81 +233,57 @@ impl Device {
...
+        }
     }
 
     /// Return all available GPU devices of supported vendors.
     pub fn all() -> Vec<&'static Device> {
         Self::all_iter().collect()
     }
 
-    pub fn by_bus_id(bus_id: BusId) -> GPUResult<&'static Device> {
+    pub fn by_pci_id(pci_id: PciId) -> GPUResult<&'static Device> {
```
Comment (dignifiedquire) [inline src/opencl/mod.rs:257]: commenet

File: src/opencl/mod.rs [hunk 10]
```diff
@@ -263,14 +372,24 @@ impl Program {
         })
     }
 
-    pub fn create_kernel(&self, name: &str, gws: usize, lws: usize) -> GPUResult<Kernel> {
+    /// Returns a kernel.
+    ///
+    /// The `global_work_size` does *not* follow the OpenCL definition. It is *not* the total
+    /// number of threads. Instead it follows CUDA's definition and is the number of
+    /// `local_work_size` sized thread groups.
+    pub fn create_kernel(
```
Comment (dignifiedquire) [inline src/opencl/mod.rs:380]: best to add a note, how the local and global are used to calculate the total number in the opencl case

File: src/opencl/utils.rs [hunk 11]
```diff
@@ -1,19 +1,45 @@
...
 
-fn get_bus_id(d: &opencl3::device::Device) -> Result<u32, GPUError> {
-    let vendor = Vendor::try_from(d.vendor()?.as_str())?;
-    match vendor {
-        Vendor::Amd => d.pci_bus_id_amd().map_err(Into::into),
-        Vendor::Nvidia => d.pci_bus_id_nv().map_err(Into::into),
-    }
+/// The PCI-ID is the combination of the PCI Bus ID and PCI Device ID.
+///
+/// It is the first two identifiers of e.g. `lcpci`:
```
Comment (dignifiedquire) [inline src/opencl/utils.rs:12]: typo

File: src/opencl/utils.rs [hunk 12]
```diff
@@ -1,19 +1,45 @@
...
+        }
+    };
+    Ok(id.into())
+}
+
+fn get_uuid(device: &opencl3::device::Device) -> GPUResult<DeviceUuid> {
+    let uuid: [u8; CL_UUID_SIZE_KHR] = device
+        .uuid_khr()?
+        .try_into()
+        .expect("opencl3 returned an invalid UUID");
```
Comment (dignifiedquire) [inline src/opencl/utils.rs:41]: panic bad, are you sure this never happens?

File: src/opencl/utils.rs [hunk 13]
```diff
@@ -54,6 +71,14 @@ fn build_device_list() -> Vec<Device> {
     let mut all_devices = Vec::new();
     let platforms: Vec<_> = opencl3::platform::get_platforms().unwrap_or_default();
 
+    // Laptops might have an integrated GPU. Such devices might have neither a PCI-ID, nor a UUID.
+    // As those devices are used for development and not for production use, it's good enough to
+    // provide a workaround which doesn't add much complexity to the code. We use a fake PCI-ID
+    // instead, which is generated by enumerating the available devices. In order to make that
+    // case easier to spot when debugging issues, a starting number which is pleasant to the human
+    // eye was choosen, that works both, decimal and hexadecimal (4660 == 0x1234).
+    let mut enumerated_device: u16 = 4660;
```
Comment (dignifiedquire) [inline src/opencl/utils.rs:80]: what if both are the case, the code should probably check at least for conflicts with any existing PCI-ID

**Review by dignifiedquire — CHANGES_REQUESTED** [review]

**Review by vmx — COMMENTED** [review]
I've also addressed the comments that were meant for that other PR.

File: src/opencl/mod.rs (hunk 4 above)
Comment (dignifiedquire) [reply src/opencl/mod.rs:115]: yeah that’s what I meant,

**Comment by vmx** [comment]
@dignifiedquire I've rebased that one, it should now be way easier to review as it actually is quite a small change :)

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**